IntegrationTest/ExerciseEndpointsTest.cs

```diff
@@ -0,0 +1,576 @@
+        solutionRepoSub.InsertSolvedRelation(Arg.Any<int>(), Arg.Any<int>()).Returns(true);
+
+        var userId = 1;
+        var roles = new List<Roles> { Roles.AnonymousUser };
+        _client.AddRoleAuth(userId, roles);
+        var requestBody = new SubmitSolutionDto(1, "x + y");
+        var jsonBody = new StringContent(
+            System.Text.Json.JsonSerializer.Serialize(requestBody),
+            Encoding.UTF8,
+            "application/json"
+        );
+
+        var response = await _client.PostAsync("/v1/exercises/1/submission", jsonBody);
+
+        Assert.Equal(HttpStatusCode.InternalServerError, response.StatusCode);
+    }
+
+    [Fact]
+    public async Task SubmitSolutionProposal_NoAuthorization_ShouldReturn_401()
+    {
+        using var scope = _factory.Services.CreateScope();
+        var solutionRepoSub = scope.ServiceProvider.GetService<ISolutionRepository>();
+        var haskellServiceSub = scope.ServiceProvider.GetService<IHaskellService>();
+
+        solutionRepoSub.CheckAnonUserExistsInSessionAsync(Arg.Any<int>(), Arg.Any<int>()).Returns(true);
+        var testcasesResponse = new List<Testcase> { new Testcase { TestCaseId = 1, IsPublicVisible = true, Input = { new TestParameter { ParameterType = "int", ParameterValue = "1" } }, Output = { new TestParameter { ParameterType = "int", ParameterValue = "1" } } } };
+        solutionRepoSub.GetTestCasesByExerciseIdAsync(Arg.Any<int>()).Returns(testcasesResponse);
+        var solutionRunnerResponse = new SolutionRunnerResponse { Action = ResponseCode.Pass };
+        haskellServiceSub.SubmitSubmission(Arg.Any<SubmissionDto>()).Returns(solutionRunnerResponse);
+        solutionRepoSub.InsertSolvedRelation(Arg.Any<int>(), Arg.Any<int>()).Returns(true);
+
+        var requestBody = new SubmitSolutionDto(1, "x + y");
+        var jsonBody = new StringContent(
+            System.Text.Json.JsonSerializer.Serialize(requestBody),
+            Encoding.UTF8,
+            "application/json"
+        );
+
+        var response = await _client.PostAsync("/v1/exercises/1/submission", jsonBody);
+
+        Assert.Equal(HttpStatusCode.Unauthorized, response.StatusCode);
+    }
+
+    private StringContent CreateExerciseRequestBody()
+    {
+        var requestBody = new
+        {
+            Name = "Sum of Two Numbers",
+            Description = "A function that calculates the sum of two integers.",
+            Solution = "solution x =\n  if x < 0\n     x * (-1)\n    else x",
+            InputParameterType = new[] { "int" },
+            OutputParamaterType = new[] { "int" },
+            Testcases = new[]
+            {
+                new { InputParams = new[] { "1" }, OutputParams = new[] { "1" }, PublicVisible = true },
+                new { InputParams = new[] { "-1" }, OutputParams = new[] { "1" }, PublicVisible = false }
+            }
+        };
+
+        return new StringContent(
```

Comment:
Good idea using a helper for the request body if it is used multiple times, however if you want you can avoid manually serializing the object and just use the object by changing _client.PostAsync -> _client.PostAsJsonAsync in the post functions.